Enable look-up by secondary index in cache

derekwaynecarr opened the pull request:
This is an enhancement to our cache to let you optionally create a store that can build a secondary index on its data.  The main use case is I want to have a store that holds a bunch of resources across namespaces, and I want to be able to then interface with that store by asking to retrieve only the resources in namespace X.

If folks are good with this pattern, I will adopt it in the admission control handlers for resource quota and limit range.

@smarterclayton @ironcladlou @erictune

derekwaynecarr committed:
Enable look-up by secondary index in cache

ironcladlou commented on pkg/client/cache/store.go:
```diff
@@ -76,6 +100,53 @@ func (c *cache) Add(obj interface{}) error {
+// it is intended to be called from a function that already has a lock on the cache
+func (c *cache) updateIndex(obj interface{}) error {
+	if c.indexFunc == nil {
+		return nil
```
Since Cache claims to implements Index, I feel like a nil `indexFunc` would be an illegal state resulting in silently broken indexing behavior.

derekwaynecarr replied:
I could see that.  I guess I could have a default indexFunc that maps to the empty string.

ironcladlou commented on pkg/client/cache/store.go:
```diff
@@ -76,6 +100,53 @@ func (c *cache) Add(obj interface{}) error {
+// it is intended to be called from a function that already has a lock on the cache
+func (c *cache) deleteFromIndex(obj interface{}) error {
+	if c.indexFunc == nil {
+		return nil
```
See [previous comment](https://github.com/GoogleCloudPlatform/kubernetes/pull/4125/files#r24119418).

ironcladlou commented on pkg/client/cache/store_test.go:
```diff
@@ -86,10 +86,47 @@ func doTestStore(t *testing.T, store Store) {
+	}
+
+	// Test Index
+	expected := map[string]util.StringSet{}
```
Nit: inline declaration?

ironcladlou commented:
Really cool idea. I made some comments inline.

derekwaynecarr commented:
@ironcladlou thanks for the feedback, will update per your comments.

after speaking with @markturansky , we thought it would be useful to support multiple index funcs.  I am going to take another pass at this work with that concept in mind.

markturansky commented:
The single index function could also be a composite of index functions.

derekwaynecarr committed:
Support multiple index functions, address feedback

derekwaynecarr commented:
Added support for giving multiple index functions, and then ability to get an index by its name.

Addressed @ironcladlou feedback to make sure cache works fine as both an Indexer and a Store.

Please review and provide any further feedback.

deads2k commented on pkg/client/cache/store.go:
```diff
@@ -76,6 +81,57 @@ func (c *cache) Add(obj interface{}) error {
 	c.lock.Lock()
 	defer c.lock.Unlock()
 	c.items[key] = obj
+	c.updateIndices(obj)
```
Can you build the list of what you need to modify in the indexes separately from actually making the update?  It would help limit contention on the lock.  Same reason the key is calculated before lock acquisition.

smarterclayton replied:
I don't think rebuilding the entire index is going to scale very long.  Why not start with incremental?  Get prev object, gen index entries for it, delete those, gen index entries for new object, insert those.


> On Feb 6, 2015, at 11:08 AM, David Eads <notifications@github.com> wrote:
> 
> In pkg/client/cache/store.go:
> 
> > @@ -76,6 +81,57 @@ func (c *cache) Add(obj interface{}) error {
> >  	c.lock.Lock()
> >  	defer c.lock.Unlock()
> >  	c.items[key] = obj
> > +	c.updateIndices(obj)
> Can you build the list of what you need to modify in the indexes separately from actually making the update? It would help limit contention on the lock. Same reason the key is calculated before lock acquisition.
> 
> —
> Reply to this email directly or view it on GitHub.
>

derekwaynecarr replied:
@smarterclayton - its not rebuilding the entire index.  your point on prev object is needed, will update.

derekwaynecarr replied:
basically, i need a diff call on add vs update.

derekwaynecarr replied:
so if on update, i get the previous object based on its key, and then remove that object from all indices, then for updated object treat like an add, it should be fine.  will make an update this afternoon.

smarterclayton replied:
Right.

----- Original Message -----
> > @@ -76,6 +81,57 @@ func (c *cache) Add(obj interface{}) error {
> >  	c.lock.Lock()
> >  	defer c.lock.Unlock()
> >  	c.items[key] = obj
> > +	c.updateIndices(obj)
> 
> so if on update, i get the previous object based on its key, and then remove
> that object from all indices, then for updated object treat like an add, it
> should be fine.  will make an update this afternoon.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/4125/files#r24257444

smarterclayton replied:
You might have to hold *a* lock while doing that (or the Add has to fail fast if there's a conflict and you have to try again)

----- Original Message -----
> > @@ -76,6 +81,57 @@ func (c *cache) Add(obj interface{}) error {
> >  	c.lock.Lock()
> >  	defer c.lock.Unlock()
> >  	c.items[key] = obj
> > +	c.updateIndices(obj)
> 
> so if on update, i get the previous object based on its key, and then remove
> that object from all indices, then for updated object treat like an add, it
> should be fine.  will make an update this afternoon.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/4125/files#r24257444

deads2k commented on pkg/client/cache/store.go:
```diff
@@ -150,10 +232,22 @@ func (c *cache) Replace(list []interface{}) error {
 	c.lock.Lock()
 	defer c.lock.Unlock()
 	c.items = items
+
```
Can you build the index before the lock and then swap it in to limit contention?  (Similar to above)

derekwaynecarr committed:
Make sure we remove old object from index before adding new

derekwaynecarr commented:
Updated to ensure that when updating an object in the indices that it handles removal before addition.  Current indexing is incremental.  An obvious solution to limiting lock contention is not there for me right now because I need the oldObject in the context of a lock in order to do removal.  I think this is a good first step and should not block the PR.

derekwaynecarr commented:
Any other comments or issues with this PR?

smarterclayton commented:
LGTM

erictune commented:
LGTM

erictune commented:
will merge this tomorrow if the oncall doesn't get to it first.
